File: 09-Graph/04-Graph-Traversal/bfs_dfs.py
from collections import deque

n = int(input("Enter number of vertices: "))
e = int(input("Enter number of edges: "))

graph = [[] for _ in range(n)]

print("Enter edges (u v):")

for _ in range(e):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

start = int(input("Enter starting vertex: "))


def bfs(start):
    visited = [False] * n
    queue = deque([start])

    visited[start] = True

    print("BFS:", end=" ")

    while queue:
        node = queue.popleft()
        print(node, end=" ")

        for neighbor in graph[node]:
            if not visited[neighbor]:
                visited[neighbor] = True
                queue.append(neighbor)

    print()


def dfs(node, visited):
    visited[node] = True
    print(node, end=" ")

    for neighbor in graph[node]:
        if not visited[neighbor]:
            dfs(neighbor, visited)


bfs(start)

visited = [False] * n

print("DFS:", end=" ")
dfs(start, visited)
print()

File: 09-Graph/04-Graph-Traversal/bfs_dfs.c
#include <stdio.h>

#define MAX 100

int graph[MAX][MAX];
int n;

void bfs(int start) {
    int visited[MAX] = {0};
    int queue[MAX];
    int front = 0, rear = 0;

    queue[rear++] = start;
    visited[start] = 1;

    printf("BFS: ");

    while(front < rear) {
        int node = queue[front++];

        printf("%d ", node);

        for(int i = 0; i < n; i++) {
            if(graph[node][i] && !visited[i]) {
                visited[i] = 1;
                queue[rear++] = i;
            }
        }
    }

    printf("\n");
}

void dfs(int node, int visited[]) {
    visited[node] = 1;

    printf("%d ", node);

    for(int i = 0; i < n; i++) {
        if(graph[node][i] && !visited[i]) {
            dfs(i, visited);
        }
    }
}

int main() {
    int e;

    printf("Enter number of vertices: ");
    scanf("%d", &n);

    printf("Enter number of edges: ");
    scanf("%d", &e);

    printf("Enter edges (u v):\n");

    for(int i = 0; i < e; i++) {
        int u, v;
        scanf("%d %d", &u, &v);

        graph[u][v] = 1;
        graph[v][u] = 1;
    }

    int start;

    printf("Enter starting vertex: ");
    scanf("%d", &start);

    bfs(start);

    int visited[MAX] = {0};

    printf("DFS: ");
    dfs(start, visited);

    printf("\n");

    return 0;
}